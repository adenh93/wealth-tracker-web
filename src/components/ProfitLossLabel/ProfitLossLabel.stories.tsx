import ProfitLossLabel, { ProfitLossLabelProps } from './ProfitLossLabel'
import { Story } from '@storybook/react'

const config = {
  title: 'Components/ProfitLossLabel',
  component: ProfitLossLabel,
}

const Template: Story<ProfitLossLabelProps> = (args) => (
  <ProfitLossLabel {...args} />
)

export const Primary = Template.bind({})

Primary.args = {
  value: 2190.5,
  type: 'currency',
}

export default config
