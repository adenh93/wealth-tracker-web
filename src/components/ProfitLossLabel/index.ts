export { default } from './ProfitLossLabel'
export type { ProfitLossLabelProps } from './ProfitLossLabel'