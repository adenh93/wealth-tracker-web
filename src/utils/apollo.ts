import { ApolloError } from '@apollo/client'

export const extractInputErrorField = <FormType>(
  error: ApolloError
): keyof FormType | null => {
  if (error.graphQLErrors.length)
    return error.graphQLErrors[0].extensions?.field as keyof FormType
  return null
}
