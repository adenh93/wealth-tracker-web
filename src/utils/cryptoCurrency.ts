import { CMC_LOGO_URL } from '../constants'

export const getCMCLogoUrl = (id: number): string => `${CMC_LOGO_URL}/${id}.png`
