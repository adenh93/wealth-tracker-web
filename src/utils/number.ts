const defaultOpts: Intl.NumberFormatOptions = {
  maximumFractionDigits: 2,
  notation: 'standard',
  signDisplay: 'never',
}

export const formatNumber = (
  value: number,
  opts: Intl.NumberFormatOptions = defaultOpts
): string =>
  new Intl.NumberFormat('en-US', {
    ...opts,
  }).format(value)

export const formatCurrency = (
  value: number,
  opts: Intl.NumberFormatOptions = defaultOpts
): string =>
  formatNumber(value, { ...opts, style: 'currency', currency: 'AUD' })

export const formatPercent = (
  value: number,
  opts: Intl.NumberFormatOptions = defaultOpts
): string => formatNumber(value, { ...opts, style: 'percent' })
