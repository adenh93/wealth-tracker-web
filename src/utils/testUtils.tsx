import { ReactElement } from 'react'
import { ThemeProvider } from '@mui/material'
import { render as tlRender, RenderResult } from '@testing-library/react'
import theme from '../theme'

export const render = (component: ReactElement): RenderResult =>
  tlRender(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
