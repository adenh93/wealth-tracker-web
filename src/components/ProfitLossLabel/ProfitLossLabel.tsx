import { FC } from 'react'
import { Typography, TypographyProps } from '@mui/material'
import { formatCurrency, formatPercent } from '../../utils/number'
import { useTheme } from '@mui/system'

export type Type = 'currency' | 'percent'

export interface ProfitLossLabelProps extends TypographyProps {
  value: number
  type: Type
}

const ProfitLossLabel: FC<ProfitLossLabelProps> = ({
  value,
  type,
  variant = 'body1',
}) => {
  const theme = useTheme()
  const isProfit: boolean = value > 0

  const getText = () =>
    type === 'currency'
      ? formatCurrency(value, {
          signDisplay: 'never',
          maximumFractionDigits: 2,
        })
      : formatPercent(value, { signDisplay: 'never', maximumFractionDigits: 2 })

  return (
    <Typography
      variant={variant}
      component="p"
      color={isProfit ? theme.palette.success.main : theme.palette.error.main}
    >
      {isProfit ? '+' : '-'} {getText()}
    </Typography>
  )
}

export default ProfitLossLabel
