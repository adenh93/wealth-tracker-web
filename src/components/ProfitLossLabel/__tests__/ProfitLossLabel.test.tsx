import ProfitLossLabel from '..'
import { render } from '../../../utils/testUtils'

describe('ProfitLossLabel component tests', () => {
  test('matches snapshot', () => {
    const { asFragment } = render(
      <ProfitLossLabel value={34560.5} type="currency" />
    )

    expect(asFragment).toMatchSnapshot()
  })

  test('it shows positive currency', () => {
    const { getByText } = render(
      <ProfitLossLabel value={34560.5} type="currency" />
    )

    expect(getByText('+ A$34,560.50')).not.toBeNull()
  })

  test('it shows negative currency', () => {
    const { getByText } = render(
      <ProfitLossLabel value={-34560.5} type="currency" />
    )

    expect(getByText('- A$34,560.50')).not.toBeNull()
  })

  test('it shows positive percent', () => {
    const { getByText } = render(
      <ProfitLossLabel value={0.052} type="percent" />
    )

    expect(getByText('+ 5.2%')).not.toBeNull()
  })

  test('it shows negative percent', () => {
    const { getByText } = render(
      <ProfitLossLabel value={-0.052} type="percent" />
    )

    expect(getByText('- 5.2%')).not.toBeNull()
  })
})
