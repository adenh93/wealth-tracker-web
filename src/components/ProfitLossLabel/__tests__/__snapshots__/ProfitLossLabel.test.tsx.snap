// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfitLossLabel component tests matches snapshot 1`] = `[Function]`;
